solr/core/src/java/org/apache/solr/core/CoreContainer.java

```diff
@@ -1259,6 +1277,20 @@ public SolrCore create(String coreName, Path instancePath, Map<String, String> p
     }
   }
 
+  /**
+   * Checks that the given path is relative to SOLR_HOME, SOLR_DATA_HOME, coreRootDirectory or one of the paths
+   * specified in solr.xml's allowPaths element.
+   * @param path path to check
+   * @throws SolrException if path is outside allowed paths
+   */
+  public void assertPathAllowed(Path path) throws SolrException {
+    if (path.normalize().equals(path) && !path.isAbsolute()) return;
```

Comment:
This doesn't cover the case of a `../foo` path, right? is that covered somewhere else?

Reply:
You are right. We need a more thorough check
* Disallow relative paths starting with "."
* Always `normalize()` the path before `toAbsolutePath()` to catch the `/var/solr/../../etc` case, else that example would return true for `startsWith("/var/solr")`

Reply:
I pushed a commit addressing this. Quite a few changes, but we now detect `..` specifically.